lib/audit.js

```diff
@@ -0,0 +1,120 @@
+const hypercoreCrypto = require('hypercore-crypto')
+const flat = require('flat-tree')
+const c = require('compact-encoding')
+const b4a = require('b4a')
+
+// this is optimised for speed over mem atm
+// can be tweaked in the future
+
+module.exports = async function auditCore (core) {
+  const corrections = {
+    tree: 0,
+    blocks: 0
+  }
+
+  const length = core.header.tree.length
+
+  const data = await readFullStorage(core.blocks.storage)
+  const tree = await readFullStorage(core.tree.storage)
+
+  const valid = new Uint8Array(Math.ceil(tree.byteLength / 40))
+  const stack = []
+
+  for (const r of core.tree.roots) {
+    valid[r.index] = 1
+    stack.push(r)
+  }
+
+  while (stack.length > 0) {
+    const node = stack.pop()
+    if ((node.index & 1) === 0) continue
+
+    const [left, right] = flat.children(node.index)
+    const leftNode = getNode(left)
+    const rightNode = getNode(right)
+
+    stack.push(leftNode, rightNode)
+
+    if (valid[node.index]) {
+      const hash = hypercoreCrypto.parent(leftNode, rightNode)
+      if (b4a.equals(hash, node.hash) && node.size === (leftNode.size + rightNode.size)) {
+        valid[leftNode.index] = 1
+        valid[rightNode.index] = 1
+        continue
+      }
+    }
+
+    if (leftNode.size) clearNode(leftNode)
+    if (rightNode.size) clearNode(rightNode)
+  }
+
+  if (corrections.tree) {
+    core.tree.cache.clear()
+  }
+
+  let i = 0
+  let nextOffset = -1
+  while (i < length) {
+    const has = core.bitfield.get(i)
+
+    if (!has) {
+      if (i + 1 === length) break
+      i = core.bitfield.findFirst(true, i + 1)
+      if (i < 0) break
+      nextOffset = -1
+      continue
+    }
+
+    if (nextOffset === -1) {
+      try {
+        nextOffset = await core.tree.byteOffset(i * 2)
+      } catch {
+        core._setBitfield(i, false)
```

Comment:
If a particular error is expected, it's more conservative to check for it explicitly, to avoid any risk of removing a bitfield that was fine.

Reply:
it should never throw basically (outside storage failure)

Reply:
Yeah, that's the only edge case I could immediately think of besides a syntax error sneaking into `byteOffset` undetected by the tests somehow.

If it only throws on edge cases, any sense in removing the try-catch and just making it crash on error? But fine with as-is solution too, it does seem very unlikely for the storage to fail just here

Reply:
(it will also throw on corruption so thats why its there)